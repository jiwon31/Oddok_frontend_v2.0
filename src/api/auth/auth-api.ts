import axios from "axios";
import { kakaoConfig } from "api/auth/kakao";

export default class AuthApi {
  static JWT_EXPIRY_TIME = 6 * 3600 * 1000; // JWT AccessToken 만료시간 (6시간)

  async login(code: string): Promise<void> {
    const token = await this.getKakaoToken(code);
    const accessToken = await this.getAuthToken(token);
    this.onLoginSuccess(accessToken);
  }

  private async getKakaoToken(code: string): Promise<string> {
    const response = await axios.post(
      "https://kauth.kakao.com/oauth/token",
      this.formUrlEncoded({
        grant_type: "authorization_code",
        client_id: kakaoConfig.clientId!,
        redirect_uri: kakaoConfig.redirectURL!,
        code,
        client_secret: kakaoConfig.clientSecret!,
      }),
      {
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
        },
      },
    );
    return response.data.access_token;
  }

  private async getAuthToken(token: string): Promise<string> {
    axios.defaults.withCredentials = true; // refreshToken을 쿠키로 받기 위해 설정
    const response = await axios.get(`/auth?token=${token}`);
    return response.data.accessToken;
  }

  private formUrlEncoded(x: Record<string, string>): string {
    return Object.keys(x).reduce((p, c) => `${p}&${c}=${encodeURIComponent(x[c]!)}`, "");
  }

  private onLoginSuccess(accessToken: string) {
    axios.defaults.headers.common["Authorization"] = `Bearer ${accessToken}`;
    setTimeout(this.getNewAccessToken, AuthApi.JWT_EXPIRY_TIME - 60000); // 토큰 만료되기 1분 전에 새로운 토큰 발급 요청
  }

  async getNewAccessToken(): Promise<void> {
    const response = await axios.get("/auth/refresh", {
      withCredentials: true,
    });
    this.onLoginSuccess(response.data.accessToken);
  }

  async logout(): Promise<void> {
    axios.get("/auth/logout");
  }

  async deleteAccount(): Promise<void> {
    axios.get("/auth/leave");
  }
}
