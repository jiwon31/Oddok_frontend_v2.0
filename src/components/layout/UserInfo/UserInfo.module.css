.user_info {
  position: relative;
}

.profile {
  display: flex;
  align-items: center;
}

.nickname {
  margin-left: 0.5rem;
  font-size: 0.85rem;
}

.info_buttons {
  position: absolute;
  width: 8rem;
  margin-top: 0.6rem;
  background-color: var(--gray900);
}

.button {
  width: 100%;
  padding: 0.8rem 0;
  font-size: 0.8rem;
}

.button:hover {
  background-color: rgba(52, 58, 64, 0.5);
}

.logout {
  border-top: 1px var(--gray700) solid;
}
