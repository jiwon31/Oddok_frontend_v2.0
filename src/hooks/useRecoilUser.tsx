import { useRecoilState } from "recoil";
import { userState } from "recoil/user-state";

export default function useRecoilUser() {
  const [user, setUser] = useRecoilState(userState);

  return { user, setUser };
}
