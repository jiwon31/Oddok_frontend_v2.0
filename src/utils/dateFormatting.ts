// Date 객체를 yyyy-mm-dd 형식의 문자열로 반환
export default function dateFormatting(date: Date): string {
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const day = date.getDate();
  return `${year}-${month < 10 ? `0${month}` : month}-${day < 10 ? `0${day}` : day}`;
}
