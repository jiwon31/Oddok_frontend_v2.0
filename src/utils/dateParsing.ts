// yyyy-mm-dd 형태의 문자열을 Date 객체로 변환
export default function dateParsing(value: string): Date {
  const date = new Date(value); // 로컬 타임존이 반영된 시간
  const offset = date.getTimezoneOffset() * 60000;
  return new Date(date.getTime() + offset);
}
