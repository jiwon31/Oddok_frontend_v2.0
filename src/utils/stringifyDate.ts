export default function stringifyDate(date: Date): string {
  const offset = date.getTimezoneOffset() * 60000;
  return new Date(+date - offset).toISOString();
}
