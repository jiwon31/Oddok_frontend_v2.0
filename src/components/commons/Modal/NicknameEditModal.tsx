import { useState, useRef } from "react";
import useRecoilUser from "hooks/useRecoilUser";
import { Modal, Input } from "components/commons";
import { useInput } from "hooks";
import useNickname from "hooks/useNickname";
import useToast from "hooks/useToast";
import styles from "./NicknameEditModal.module.css";

export default function NicknameEditModal({ onClose }: { onClose: () => void }) {
  const { user } = useRecoilUser();
  const [nickname, setNickname] = useState(user?.nickname);
  const inputRef = useRef<HTMLInputElement>(null);
  const [isDisabled, setIsDisabled] = useState(false);
  const { updateNickname } = useNickname();
  const { successToast } = useToast();

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (!e.target.value) {
      setIsDisabled(true);
    } else if (e.target.value && isDisabled) {
      setIsDisabled(false);
    }
    setNickname(e.target.value);
  };

  const editNickname = async (name: string) =>
    updateNickname.mutate(name, {
      onSuccess: () => {
        successToast("닉네임 수정이 완료되었습니다.");
        onClose();
      },
    });

  const { pressEnter } = useInput(inputRef, () => editNickname(nickname!), isDisabled);

  const content = (
    <label htmlFor="nickname">
      <p className={styles.content}>닉네임</p>
      <Input ref={inputRef} value={nickname} maxLength={8} onChange={handleChange} onKeyPress={pressEnter} />
    </label>
  );

  return (
    <Modal
      title="닉네임 수정"
      content={content}
      onClose={onClose}
      onAction={{
        text: "확인",
        action: () => editNickname(nickname!),
      }}
      disabled={isDisabled}
    />
  );
}
