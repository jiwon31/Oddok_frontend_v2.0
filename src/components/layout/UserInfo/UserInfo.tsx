import { useRef, useState } from "react";
import { useRecoilValue } from "recoil";
import { userState } from "recoil/user-state";
import { NicknameEditModal } from "components/commons";
import { KAKAO_LOGOUT_URL } from "api/auth/kakao";
import { useModal, useOutSideClick } from "hooks";
import { Profile } from "assets/icons";
import styles from "./UserInfo.module.css";

export default function UserInfo() {
  const user = useRecoilValue(userState)!;
  const [isDropdown, setIsDropdown] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);
  const { isModal, openModal, closeModal } = useModal();

  const handleProfileClick = () => setIsDropdown((prev) => !prev);
  const handleNicknameEditBtnClick = () => {
    openModal();
    setIsDropdown(false);
  };

  useOutSideClick(dropdownRef, () => setIsDropdown(false));

  return (
    <>
      {isModal && <NicknameEditModal onClose={closeModal} />}
      <div className={styles.user_info} ref={dropdownRef}>
        <button type="button" className={styles.profile} onClick={handleProfileClick}>
          <Profile />
          <span className={styles.nickname}>{user.nickname}</span>
        </button>
        {isDropdown && (
          <ul className={styles.info_buttons}>
            <li>
              <button type="button" className={styles.button} onClick={handleNicknameEditBtnClick}>
                닉네임 수정
              </button>
            </li>
            <li>
              <a href={KAKAO_LOGOUT_URL}>
                <button type="button" className={`${styles.button} ${styles.logout}`}>
                  로그아웃
                </button>
              </a>
            </li>
          </ul>
        )}
      </div>
    </>
  );
}
