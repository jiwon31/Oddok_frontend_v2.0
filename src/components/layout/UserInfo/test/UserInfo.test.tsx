import useRecoilUser from "hooks/useRecoilUser";
import NicknameEditModal from "components/commons/Modal/NicknameEditModal";
import { withRecoil } from "tests/utils";
import renderer from "react-test-renderer";
import { render, screen, waitForElementToBeRemoved } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { fakeUser } from "tests/user";
import { KAKAO_LOGOUT_URL } from "api/auth/kakao";
import UserInfo from "../UserInfo";

jest.mock("hooks/useRecoilUser");
jest.mock("components/commons/Modal/NicknameEditModal");
const useRecoilUserMock = useRecoilUser as jest.MockedFunction<typeof useRecoilUser>;
const NicknameEditModalMock = NicknameEditModal as jest.MockedFunction<typeof NicknameEditModal>;

const UserInfoComponent = withRecoil(<UserInfo />);

describe("UserInfo", () => {
  beforeEach(() => {
    useRecoilUserMock.mockImplementation(() => ({ user: fakeUser, setUser: jest.fn() }));
  });

  afterEach(() => {
    useRecoilUserMock.mockReset();
    NicknameEditModalMock.mockReset();
  });

  it("renders", () => {
    const component = renderer.create(UserInfoComponent);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("dropdown should be toggled when info button is clicked", async () => {
    render(UserInfoComponent);

    userEvent.click(screen.getByRole("button"));
    expect(await screen.findByRole("list")).toBeInTheDocument();

    userEvent.click(screen.getAllByRole("button")[0]!);
    await waitForElementToBeRemoved(screen.queryByRole("list"));
    expect(screen.queryByRole("list")).toBeNull();
  });

  it("should pops up modal when edit nickname button is clicked", async () => {
    NicknameEditModalMock.mockImplementation(() => <p>닉네임 수정 모달</p>);
    render(UserInfoComponent);

    userEvent.click(screen.getByRole("button"));
    await screen.findByRole("list");
    userEvent.click(screen.getByText(/닉네임 수정/i));
    expect(await screen.findByText("닉네임 수정 모달")).toBeInTheDocument();
  });

  it("should go to logout page when logout button is clicked", async () => {
    render(UserInfoComponent);

    userEvent.click(screen.getByRole("button"));
    await screen.findByRole("list");
    expect(screen.getByRole("link", { name: "로그아웃" })).toHaveAttribute("href", KAKAO_LOGOUT_URL);
  });
});
