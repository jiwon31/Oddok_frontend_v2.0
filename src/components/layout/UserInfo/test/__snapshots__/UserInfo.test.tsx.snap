// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfo renders 1`] = `
<div
  className="user_info"
>
  <button
    className="profile"
    onClick={[Function]}
    type="button"
  >
    <svg>
      person-circle.svg
    </svg>
    <span
      className="nickname"
    >
      user
    </span>
  </button>
</div>
`;
